synapse/push/pusherpool.py

```diff
@@ -152,6 +152,16 @@ async def add_pusher(
             )
         )
 
+        # Before we actually create the pusher, we check if the user already has one for
+        # this app ID and pushkey. If so, we want to keep the access token in place,
+        # since this could be one device modifying (e.g. enabling/disabling) another
+        # device's pusher.
+        existing_config = await self._get_pusher_config_for_user_by_app_id_and_pushkey(
+            user_id, app_id, pushkey
+        )
+        if existing_config:
+            access_token = existing_config.access_token
+
```

Comment:
Wonder if that needs a test case---I didn't catch this at all :(

I also wonder if the method name `add_pusher` is possibly misleading. It actually does an upsert in the datastore! Should this be `upsert_pusher`? Or `add_or_update_pusher?`

Reply:
Good call on both points. `add_or_update_pusher` sounds more relevant since `upsert` sounds purely database-related to me.